src/mono/mono/metadata/assembly-internals.h

```diff
@@ -141,6 +141,30 @@ mono_assembly_get_image_internal (MonoAssembly *assembly);
 void
 mono_set_assemblies_path_direct (char **path);
 
+typedef enum {
+	BUNDLED_BLOB,
+	BUNDLED_ASSEMBLY,
+	BUNDLED_SATELLITE_ASSEMBLY,
+	BUNDLED_BINARY,
+	BUNDLED_RESOURCE_COUNT,
+} BundledResourceType;
+
+typedef struct BundledResource {
+	const char *culture; // Satellite assemblies
+	const unsigned char *data;
+	unsigned int size;
+	BundledResourceType type;
+} BundledResource;
+
+void
+mono_add_bundled_resource (const char *name, const char *culture, const unsigned char *data, unsigned int size, BundledResourceType type);
+
+void
+mono_get_bundled_resource_data (const char *name, const unsigned char **out_data, unsigned int *out_size);
+
+void
+mono_register_bundled_resources (void);
```

Comment:
If this method gets called, would attempts to add assemblies and satellite assemblies using mono_add_bundled_resource be incorrect? I assume that is the case since that's how the underlying API's used to work, so maybe we should assert that no new calls are done after mono_register_bundled_resoruces has been called?

Reply:
If there are subsequent calls to `mono_add_bundled_resource` after `mono_register_bundled_resources` has been invoked, I think the user would need to call `mono_register_bundled_resources` again to make sure all satellite assembly and assembly resources are properly registered.

The underlying APIs being `mono_register_bundled_satellite_assemblies` and `mono_register_bundled_assemblies`? Is it incorrect to have multiple calls to the register APIs?
I would imagine that subsequent calls to `mono_register_bundled_resources` would always be a super set of the previous call to `mono_register_bundled_resources` because the `bundled_resources` doesn't lose its entries.

The scenario in which `mono_add_bundled_resource` is called too early would be when the user building the self-contained library uses a custom runtime init callback right? I dont think there would be anything wrong with say

`mono_add_bundled_resource` x10 -> `mono_register_bundled_resources` -> `mono_add_bundled_resource` x15 -> `mono_register_bundled_resources`

It would just be that the first `mono_register_bundled_resources` was unnecessary.

Reply:
It would be a user error of invoking `mono_register_bundled_resources` too early, and asserting that no new calls to `mono_add_bundled_resource` would be punishing those early calls, instead of allowing subsequent calls to `mono_register_bundled_resources`?

In any case, I didn't think about multiple calls to `mono_register_bundled_resources`. I would need to free the memory associated with the `MonoBundledSatelliteAssembly **` and `MonoBundledAssembly **` because `mono_register_bundled_assemblies` and `mono_register_bundled_satellite_assemblies` don't free its previous value right?

Reply:
Because the new hash table is bootstrapped into preexisting bundle search logic, there is no need to invoke `mono_register_bundled_assemblies` or `mono_register_bundled_satellite_assemblies`. However, we still need to assert that the runtime is not initialized in `mono_add_bundled_resource` as suggested in https://github.com/dotnet/runtime/pull/84191#discussion_r1163003217